File: 2022/day2b_rock_paper_scissors.swift

enum Choice: Equatable {
    case rock
    case paper
    case scissors
}

func score(_ choice: Choice) -> Int {
    switch choice {
    case .rock:     return 1
    case .paper:    return 2
    case .scissors: return 3
    }
}

func getChoice(_ s: Substring) -> Choice {
    switch s {
    case "A": return .rock
    case "B": return .paper
    case "C": return .scissors
    default: unreachable()
    }
}

func unreachable() -> Never {
    fatalError("unreachable")
}

var total = 0
while let line = readLine() {
    let parts = line.split(separator: " ")
    let opponent = parts[0]
    let your = parts[1]

    let opponent_choice = getChoice(opponent)
    let your_choice: Choice

    switch (opponent_choice, your) {
    case (.rock, "X"):
        your_choice = .scissors
    case (.paper, "X"):
        your_choice = .rock
    case (.scissors, "X"):
        your_choice = .paper
    case let (a, "Y"):
        your_choice = a
    case (.rock, "Z"):
        your_choice = .paper
    case (.paper, "Z"):
        your_choice = .scissors
    case (.scissors, "Z"):
        your_choice = .rock
    default: unreachable()
    }

    switch (your) {
    case "Y":
        total += 3
    case "Z":
        total += 6
    default: break
    }

    total += score(your_choice)
}

print("\(total)")

File: 2022/day2b_rock_paper_scissors.hs
import Data.Char (isSpace)
import Control.Arrow ((>>>))
import Data.Function ((&))

data Choice = Rock | Paper | Scissors deriving Eq

instance Read Choice where
    readsPrec _ s = do choice <- case stripped of
                                   'A':_ -> pure Rock
                                   'X':_ -> pure Rock
                                   'B':_ -> pure Paper
                                   'Y':_ -> pure Paper
                                   'C':_ -> pure Scissors
                                   'Z':_ -> pure Scissors
                                   []    -> []
                       return (choice, tail stripped)
        where stripped = s & span isSpace & snd

Rock     `beats` Scissors = True
Paper    `beats` Rock     = True
Scissors `beats` Paper    = True
_        `beats` _        = False

whatBeats Rock     = Paper
whatBeats Paper    = Scissors
whatBeats Scissors = Rock

whatIsBeatenBy Paper    = Rock
whatIsBeatenBy Scissors = Paper
whatIsBeatenBy Rock     = Scissors

choiceScore Rock     = 1
choiceScore Paper    = 2
choiceScore Scissors = 3

score opponentsChoice yourChoice = choiceScore yourChoice + matchResultScore
    where matchResultScore = if yourChoice `beats` opponentsChoice
                               then 6
                               else if yourChoice == opponentsChoice
                               then 3
                               else 0

main = interact
     ( lines
   >>> fmap ( words
          >>> \[opponent, instruction] -> let opponentsChoice = read opponent
                                              yourChoice = case instruction of
                                                             "X" -> whatIsBeatenBy opponentsChoice
                                                             "Y" -> opponentsChoice
                                                             "Z" -> whatBeats opponentsChoice
                                           in score opponentsChoice yourChoice
            )
   >>> sum
   >>> show
   >>> (++"\n")
     )
